docs/source/howto/data.rst

```diff
@@ -18,7 +18,166 @@ Importing data
+    value = node.value
+
+As said before, in addition to their attributes, data types can also store their properties in the file repository.
+Imagine a data type that needs to wrap a single text file.
+
+.. code:: python
+
+    import os
+    from aiida.orm import Data
+
+
+    class TextFileData(Data):
+        """Data class that can be used to wrap a single text file by storing it in its file repository."""
+
+        def __init__(self, filepath, **kwargs):
+            """Construct a new instance and set the contents to that of the file.
+
+            :param file: an absolute filepath of the file to wrap
+            """
+            super().__init__(**kwargs)
+
+            filename = os.path.basename(filepath)  # Get the filename from the absolute path
+            self.put_object_from_file(filepath, filename)  # Store the file in the repository under the given filename
+            self.set_attribute('filename', filename)  # Store in the attributes what the filename is
+
+        def get_content(self):
+            """Return the content of the single file stored for this data node.
+
+            :return: the content of the file as a string
+            """
+            filename = self.get_attribute('filename')
+            return self.get_object_content(filename)
+
+To create a new instance of this data type and get its content:
+
+.. code:: python
+
+    node = TextFileData(filepath='/some/absolute/path/to/file.txt')
+    node.get_content()  # This will return the content of the file
+
+This example is a simplified version of the :class:`~aiida.orm.nodes.data.singlefile.SinglefileData` data class that ships with ``aiida-core``.
+If this happens to be your use case (or very close to it), it is of course better to use that class, or you can sub class it and adapt it where needed.
+
+The two new data types we have just implemented are of course trivial and not very useful, but the concepts are flexible and can easily be applied to more complex use-cases.
+The following section will provide useful guidelines on how to optimally design new data types.
+
+
+.. _how-to:data:plugin:design-guidelines:
+
+Database or repository?
+-----------------------
+
+When deciding where to store a property of a data type, one has to choose between the database and the file repository.
+The database will make it possible to search in the provenance graph based on criteria based on the property, e.g. all ``NewData`` nodes where the property is greater than 0.
+The downside is that storing too much information in the database can make it sluggish.
+Therefore, big data (think large files), whose content does not necessarily need to be queried for, is better stored in the file repository.
+Of course a data type may need to store multiple properties of varying character, but both storage locations can safely be used in parellel.
+When choosing the database as the storage location, the properties should be stored using the node *attributes*.
```

Comment:
```suggestion
Properties stored as part of the node's *attributes* are stored in the database.
```
Is that right? I'm trying to simplify the sentence, but I'm not 100% sure whether I got cause and effect right.

Reply:
Technically it is correct yes. What I think is missing here is a direct reference to the actual methods of the `Node` class. Without those, the term `attributes` is too vague I think. Maybe I can add a sentence like:
```
The node class has various methods to set its attributes, such as :py:`~aiida.orm.node.Node.set_attribute` and :py:`~aiida.orm.node.Node.set_attribute_many`
```

Reply:
Maybe you can provide an example that references those methods?